python/lsst/pex/exceptions/exceptions.cc

```diff
@@ -0,0 +1,155 @@
+#include <sstream>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/exception_translator.h>
+
+#include "lsst/pex/exceptions/Exception.h"
+#include "lsst/pex/exceptions/Runtime.h"
+
+using namespace lsst::pex::exceptions;
+
+namespace py = pybind11;
+
+// Helper function to show a Python warning when exception translation fails.
+static void tryLsstExceptionWarn(const char * message) {
+    // Try to warn that exception translation failed, if we fail, clear the exception raised by the
+    // warning attempt so we can raise a less-informative exception based on the original.
+    int s = PyErr_WarnEx(PyExc_Warning, message, 1);
+    if (s) {
+        PyErr_Clear();
+    }
+}
+
+// Raise a Python exception that wraps the given C++ exception instance.
+//
+// Most of the work is delegated to the pure-Python function pex.exceptions.wrappers.translate(),
+// which looks up the appropriate Python exception class from a dict that maps C++ exception
+// types to their custom Python wrappers.  Everything else here is basically just importing that
+// module, preparing the arguments, and calling that function, along with the very verbose error
+// handling required by the Python C API.
+//
+// If any point we fail to translate the exception, we print a Python warning and raise the built-in
+// Python RuntimeError exception with the same message as the C++ exception.
+static void raiseLsstException(py::object &pyex) {
+    PyObject* module = PyImport_ImportModule("lsst.pex.exceptions.wrappers");
+    if (!module) {
+        tryLsstExceptionWarn("Failed to import C++ Exception wrapper module.");
+    } else {
+        PyObject* translate = PyObject_GetAttrString(module, "translate");
+        if (!translate) {
+            tryLsstExceptionWarn("Failed to find translation function for C++ Exceptions.");
+        } else {
+            // Calling the Python translate() returns an instance of the appropriate Python
+            // exception that wraps the C++ exception instance that we give it.
+            PyObject* instance = PyObject_CallFunctionObjArgs(translate, pyex.ptr(), NULL);
+            if (!instance) {
+                // we actually expect a null return here, as translate() should raise an exception
+                tryLsstExceptionWarn("Failed to translate C++ Exception to Python.");
+            } else {
+                PyObject* type = PyObject_Type(instance);
+                PyErr_SetObject(type, instance);
+                Py_DECREF(type);
+                Py_DECREF(instance);
+            }
+            Py_DECREF(translate);
+        }
+        Py_DECREF(module);
```

Comment:
Agreed. I ported this (and other similar explicit Python API using code) over from the Swig solution. But will convert it to use RAII with py::object as much as possible.